<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Jasmine Chong, Jeff Xia" />

<meta name="date" content="2020-04-15" />

<title>Integrated Pathway Analysis</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Integrated Pathway Analysis</h1>
<h4 class="author">Jasmine Chong, Jeff Xia</h4>
<h4 class="date">2020-04-15</h4>



<div id="introduction" class="section level2">
<h2>1. Introduction</h2>
<p>This module performs integrated metabolic pathway analysis on results obtained from combined metabolomics and gene expression studies conducted under the same experimental conditions. This approach exploits KEGG metabolic pathway models to complete the analysis. The underlying assumption behind this module is that by combining evidence from both changes in gene expression and metabolite concentrations, one is more likely to pinpoint the pathways involved in the underlying biological processes.</p>
<p>To this end, users need to supply a list of genes and metabolites of interest that have been identified from the same samples or obtained under similar conditions. The metabolite list can be selected from the results of a previous analysis downloaded from MetaboAnalyst. Similarly, the gene list can be easily obtained using many excellent web-based tools such as GEPAS or INVEX.</p>
<p>After users have uploaded their data, the genes and metabolites are then mapped to KEGG metabolic pathways for over-representation analysis and pathway topology analysis. Topology analysis uses the structure of a given pathway to evaluate the relative importance of the genes/compounds based on their relative location. Clicking on the name of a specific pathway will generate a graphical representation of that pathway highlighted with the matched genes/metabolites. Users must keep in mind that unlike transcriptomics, where the entire transcriptome is routinely mapped, current metabolomic technologies only capture a small portion of the metabolome. This difference can lead to potentially biased results. To address this issue, the current implementation of this omic integration module allows users to explore the enriched pathways based either on joint evidence or on the evidence obtained from one particular omic platform for comparison.</p>
</div>
<div id="download-necessary-files" class="section level2">
<h2>1.1 Download Necessary Files</h2>
<p>Before beginning, <strong>two example files must be downloaded</strong> and placed in your working directory, the gene list and the metabolite list. The example data comes from an integrative analysis of the transcriptome and metabolome to identify (metabolites/genes) biomarkers of intrahepatic cholangiocarcinoma (ICC) in 16 individuals. Please download these files prior to using the tutorial below as per the R code below.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw">download.file</span>(<span class="st">&quot;https://www.metaboanalyst.ca/MetaboAnalyst/resources/data/integ_genes.txt&quot;</span>, <span class="st">&quot;integ_genes.txt&quot;</span>, <span class="st">&quot;curl&quot;</span>)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"></a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="kw">download.file</span>(<span class="st">&quot;https://www.metaboanalyst.ca/MetaboAnalyst/resources/data/integ_cmpds.txt&quot;</span>, <span class="st">&quot;integ_cmpds.txt&quot;</span>, <span class="st">&quot;curl&quot;</span>)</a></code></pre></div>
</div>
<div id="joint-integrated-pathway-analysis" class="section level2">
<h2>2. Joint (Integrated) Pathway Analysis</h2>
<p>Enrichment analysis aims to evaluate whether the observed genes and metabolites in a particular pathway are significantly enriched (appeatr more than expected by random chance) within the dataset. You can choose over-representation analysis (ORA) based on either hypergenometrics analysis or Fisher’s exact method.</p>
<p>The topology analysis aims to evaluate whether a given gene or metabolite plays an important role in a biological response based on its position within a pathway. Degree Centrality measures the number of links that connect to a node (representing either a gene or metabolite) within a pathway; Closeness Centrality measures the overall distance from a given node to all other nodes in a pathway; Betweenness Centrality measures the number of shortest paths from all nodes to all the others that pass through a given node within a pathway.</p>
<p>Users can choose one of three different modes of pathways: - the gene-metabolite mode (default, “integ”) allows joint-analysis and visualization of both significant genes and metabolites; while the gene-centric (“genetic”) or metabolite-centric (“metab”) mode allows users to identify enriched pathways driven by significant genes or metabolites, respectively.</p>
<div id="step-1-perform-genemetabolite-name-mapping" class="section level3">
<h3>2.1 Step 1: Perform Gene/Metabolite Name Mapping</h3>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw">library</span>(MetaboAnalystR)</a>
<a class="sourceLine" id="cb2-2" data-line-number="2"></a>
<a class="sourceLine" id="cb2-3" data-line-number="3"><span class="co"># Initiate MetaboAnalyst</span></a>
<a class="sourceLine" id="cb2-4" data-line-number="4">mSet&lt;-<span class="kw">InitDataObjects</span>(<span class="st">&quot;conc&quot;</span>, <span class="st">&quot;pathinteg&quot;</span>, <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb2-5" data-line-number="5"></a>
<a class="sourceLine" id="cb2-6" data-line-number="6"><span class="co"># Set organism library</span></a>
<a class="sourceLine" id="cb2-7" data-line-number="7">mSet&lt;-<span class="kw">SetOrganism</span>(mSet, <span class="st">&quot;hsa&quot;</span>)</a>
<a class="sourceLine" id="cb2-8" data-line-number="8"></a>
<a class="sourceLine" id="cb2-9" data-line-number="9"><span class="co"># Set the name of your file containing your gene list</span></a>
<a class="sourceLine" id="cb2-10" data-line-number="10">geneListFile&lt;-<span class="st">&quot;integ_genes.txt&quot;</span></a>
<a class="sourceLine" id="cb2-11" data-line-number="11"></a>
<a class="sourceLine" id="cb2-12" data-line-number="12"><span class="co"># Read in your gene list file</span></a>
<a class="sourceLine" id="cb2-13" data-line-number="13">geneList&lt;-<span class="kw">readChar</span>(geneListFile, <span class="kw">file.info</span>(geneListFile)<span class="op">$</span>size)</a>
<a class="sourceLine" id="cb2-14" data-line-number="14"></a>
<a class="sourceLine" id="cb2-15" data-line-number="15"><span class="co"># Perform gene mapping of your file</span></a>
<a class="sourceLine" id="cb2-16" data-line-number="16">mSet&lt;-<span class="kw">PerformIntegGeneMapping</span>(mSet, geneList, <span class="st">&quot;hsa&quot;</span>, <span class="st">&quot;symbol&quot;</span>);</a>
<a class="sourceLine" id="cb2-17" data-line-number="17"></a>
<a class="sourceLine" id="cb2-18" data-line-number="18"><span class="co"># Set the name of your file containing your compound list</span></a>
<a class="sourceLine" id="cb2-19" data-line-number="19">cmpdListFile&lt;-<span class="st">&quot;integ_cmpds.txt&quot;</span></a>
<a class="sourceLine" id="cb2-20" data-line-number="20"></a>
<a class="sourceLine" id="cb2-21" data-line-number="21"><span class="co"># Read in your compound list file</span></a>
<a class="sourceLine" id="cb2-22" data-line-number="22">cmpdList&lt;-<span class="kw">readChar</span>(cmpdListFile, <span class="kw">file.info</span>(cmpdListFile)<span class="op">$</span>size)</a>
<a class="sourceLine" id="cb2-23" data-line-number="23"></a>
<a class="sourceLine" id="cb2-24" data-line-number="24"><span class="co"># Perform compound mapping of your file</span></a>
<a class="sourceLine" id="cb2-25" data-line-number="25">mSet&lt;-<span class="kw">PerformIntegCmpdMapping</span>(mSet, cmpdList, <span class="st">&quot;hsa&quot;</span>, <span class="st">&quot;kegg&quot;</span>);</a>
<a class="sourceLine" id="cb2-26" data-line-number="26"></a>
<a class="sourceLine" id="cb2-27" data-line-number="27"><span class="co"># Create a mapping result table</span></a>
<a class="sourceLine" id="cb2-28" data-line-number="28">mSet&lt;-<span class="kw">CreateMappingResultTable</span>(mSet)</a>
<a class="sourceLine" id="cb2-29" data-line-number="29"></a>
<a class="sourceLine" id="cb2-30" data-line-number="30"><span class="co"># Prepare data for joint pathway analysis</span></a>
<a class="sourceLine" id="cb2-31" data-line-number="31">mSet&lt;-<span class="kw">PrepareIntegData</span>(mSet);</a></code></pre></div>
<div id="trouble-shooting" class="section level4">
<h4>Trouble Shooting</h4>
<p>When performing compound mapping (PerformIntegCmpdMapping), you may come across this error:</p>
<p>[1] “Loading files from server unsuccessful. Ensure curl is downloaded on your computer.” Error in .read.metaboanalyst.lib(“compound_db.rds”) : objet ‘my.lib’ introuvable</p>
<p>This means that the function was unable to download the “compound_db.rds” file from the MetaboAnalyst server. This could be because curl is not installed on your computer. Download from here: <a href="https://curl.haxx.se/download.html" class="uri">https://curl.haxx.se/download.html</a>. curl is a command line tool for transferring files using a URL. Once curl is installed, try the function again. If it still does not work, download the “compound_db.rds” file manually from this link: <a href="https://www.dropbox.com/s/nte1ok440bt1l8w/compound_db.rds?dl=0" class="uri">https://www.dropbox.com/s/nte1ok440bt1l8w/compound_db.rds?dl=0</a>. Make sure that the file is always in your current working directory when performing compound mapping.</p>
</div>
</div>
<div id="step-2-perform-joint-pathway-analysis" class="section level3">
<h3>2.2 Step 2: Perform Joint Pathway Analysis</h3>
<p>To perform Joint Pathway Analysis, we show two options below. Note you can only perform the analysis once per session.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="co">#### OPTION 1 ####</span></a>
<a class="sourceLine" id="cb3-2" data-line-number="2"><span class="co"># Perform integrated pathway analysis, using hypergeometric test, degree centrality, and the gene-metabolite pathways</span></a>
<a class="sourceLine" id="cb3-3" data-line-number="3"><span class="co"># Use the current integrated metabolic pathways</span></a>
<a class="sourceLine" id="cb3-4" data-line-number="4"><span class="co"># Saves the output as MetaboAnalyst_result_pathway.csv</span></a>
<a class="sourceLine" id="cb3-5" data-line-number="5">mSet&lt;-<span class="kw">PerformIntegPathwayAnalysis</span>(mSet, <span class="st">&quot;dc&quot;</span>, <span class="st">&quot;hyper&quot;</span>, <span class="st">&quot;current&quot;</span>, <span class="st">&quot;integ&quot;</span>, <span class="st">&quot;query&quot;</span>);</a>
<a class="sourceLine" id="cb3-6" data-line-number="6"></a>
<a class="sourceLine" id="cb3-7" data-line-number="7"><span class="co"># View the output of the pathway analysis</span></a>
<a class="sourceLine" id="cb3-8" data-line-number="8">mSet<span class="op">$</span>dataSet<span class="op">$</span>path.mat</a></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="co">#### OPTION 2 ####</span></a>
<a class="sourceLine" id="cb4-2" data-line-number="2"><span class="co"># Perform integrated pathway analysis using the current library of all pathways (not just metabolic)</span></a>
<a class="sourceLine" id="cb4-3" data-line-number="3"><span class="co"># Fisher's exact test for enrichment and betweenness centrality for topology measure</span></a>
<a class="sourceLine" id="cb4-4" data-line-number="4">mSet&lt;-<span class="kw">PerformIntegPathwayAnalysis</span>(mSet, <span class="st">&quot;bc&quot;</span>, <span class="st">&quot;fisher&quot;</span>, <span class="st">&quot;current&quot;</span>, <span class="st">&quot;all&quot;</span>, <span class="st">&quot;query&quot;</span>);</a></code></pre></div>
<p><strong>Pathway Analysis Overview</strong></p>
<p>The scatter plot below show a summary of the joint evidence from enrichment analysis (p-values) and topology analysis (pathway impact).</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="co"># Perform pathway analysis</span></a>
<a class="sourceLine" id="cb5-2" data-line-number="2">mSet&lt;-<span class="kw">PlotPathSummary</span>(mSet, <span class="st">&quot;path_view_0_&quot;</span>, <span class="st">&quot;png&quot;</span>, <span class="dv">72</span>, <span class="dt">width=</span><span class="ot">NA</span>)</a></code></pre></div>
</div>
</div>
<div id="sweave-report" class="section level2">
<h2>3. Sweave Report</h2>
<p>Following analysis, a comprehensive report can be generated which contains a detailed description of each step performed in the R package, embedded with graphical and tabular outputs. To prepare the sweave report, please use the <em>PreparePDFReport</em> function. You must ensure that you have the nexessary Latex libraries to generate the report (i.e. pdflatex, LaTexiT). The object created <em>must</em> be named <em>mSet</em>, and specify the user name in quotation marks.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="kw">PreparePDFReport</span>(mSet, <span class="st">&quot;My Name&quot;</span>)</a></code></pre></div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
